src/flint/types/fmpz_mpoly.pyx

```diff
@@ -135,362 +192,834 @@ cdef class fmpz_mpoly(flint_mpoly):
+                res = create_fmpz_mpoly(self.ctx)
                 fmpz_mpoly_div(res.val, (<fmpz_mpoly>self).val, (<fmpz_mpoly>other).val, res.ctx.val)
                 return res
         return NotImplemented
 
+    def __rfloordiv__(self, other):
+        cdef fmpz_mpoly res
+        if not self:
+            raise ZeroDivisionError("fmpz_mpoly division by zero")
+        other = any_as_fmpz(other)
+        if other is not NotImplemented:
+            other = fmpz_mpoly(other, self.ctx)
+            res = create_fmpz_mpoly(self.ctx)
+            fmpz_mpoly_div(res.val, (<fmpz_mpoly>other).val,  self.val, res.ctx.val)
+            return res
+        return NotImplemented
+
+    def __truediv__(self, other):
+        cdef:
+            fmpz_mpoly res
+
+        if typecheck(other, fmpz_mpoly):
+            if not other:
+                raise ZeroDivisionError("fmpz_mpoly division by zero")
+            elif self.ctx is not (<fmpz_mpoly>other).ctx:
+                raise IncompatibleContextError(f"{self.ctx} is not {(<fmpz_mpoly>other).ctx}")
+
+            res = create_fmpz_mpoly(self.ctx)
+            if fmpz_mpoly_divides(res.val, self.val, (<fmpz_mpoly>other).val, self.ctx.val):
+                return res
+            else:
+                raise DomainError("fmpz_mpoly division is not exact")
+        else:
+            o = any_as_fmpz(other)
+            if o is NotImplemented:
+                return NotImplemented
+            elif not o:
+                raise ZeroDivisionError("fmpz_mpoly division by zero")
+            res = create_fmpz_mpoly(self.ctx)
+            if fmpz_mpoly_scalar_divides_fmpz(res.val, self.val, (<fmpz>o).val, self.ctx.val):
+                return res
+            else:
+                raise DomainError("fmpz_mpoly division is not exact")
+
+    def __rtruediv__(self, other):
+        cdef fmpz_mpoly res
+        if not self:
+            raise ZeroDivisionError("fmpz_mpoly division by zero")
+        o = any_as_fmpz(other)
+        if o is NotImplemented:
+            return NotImplemented
+        res = create_fmpz_mpoly(self.ctx)
+        fmpz_mpoly_set_fmpz(res.val, (<fmpz>o).val, self.ctx.val)
+        return res / self
+
     def __mod__(self, other):
         return divmod(self, other)[1]
 
+    def __rmod__(self, other):
+        return divmod(other, self)[1]
+
+    def __call__(self, *args, **kwargs):
+        cdef:
+            fmpz_mpoly res
+            fmpz_mpoly res2
+            fmpz_mpoly_ctx res_ctx
+            fmpz_vec V
+            fmpz vres
+            fmpz_mpoly_vec C
+            slong i, nvars = self.ctx.nvars(), nargs = len(args)
+
+        if not args and not kwargs:
+            return self
+        elif args and kwargs:
+            raise ValueError("only supply positional or keyword arguments")
+        elif len(args) < nvars and not kwargs:
+            raise ValueError("partial application requires keyword arguments")
+
+        if kwargs:
```

Comment:
I think that allowing kwargs in `__call__` is overloading things too far. The interface for `__call__` should be simple. Other methods can be used for variations like partial evaluation.

Reply:
The `__call__` methods are far too complicated. Separate simpler methods are needed.

If kwargs is used for the names of the symbols like `p(x=2)` then I would prefer to avoid that because that interface restricts the symbol names to be strings but we will likely want to allow them to be hashable objects in future.